packages/edit-site/src/components/resizable-frame/index.js

```diff
@@ -33,108 +33,124 @@ const HANDLE_STYLES_OVERRIDE = {
 };
 
 const MIN_FRAME_SIZE = 340;
-const lerp = ( a, b, amount ) => {
-	return a + ( b - a ) * amount;
-};
 
-function ResizableFrame( { isFull, children } ) {
+function calculateNewHeight( width, initialAspectRatio ) {
+	const lerp = ( a, b, amount ) => {
+		return a + ( b - a ) * amount;
+	};
+
+	// Calculate the intermediate aspect ratio based on the current width.
+	const lerpFactor =
+		1 -
+		Math.max(
+			0,
+			Math.min(
+				1,
+				( width - MIN_FRAME_SIZE ) / ( 1300 - MIN_FRAME_SIZE )
+			)
+		);
+
+	// Calculate the height based on the intermediate aspect ratio
+	// ensuring the frame arrives at a 9:19.5 aspect ratio.
+	const intermediateAspectRatio = lerp(
+		initialAspectRatio,
+		9 / 19.5,
+		lerpFactor
+	);
+
+	return width / intermediateAspectRatio;
+}
+
+function ResizableFrame( { isFullWidth, children } ) {
 	const [ frameSize, setFrameSize ] = useState( {
 		width: '100%',
 		height: '100%',
 	} );
+	const [ startingWidth, setStartingWidth ] = useState();
 	const [ isResizing, setIsResizing ] = useState( false );
 	const [ isHovering, setIsHovering ] = useState( false );
 	const [ isOversized, setIsOversized ] = useState( false );
 	const { setCanvasMode } = unlock( useDispatch( editSiteStore ) );
 	const initialAspectRatioRef = useRef( null );
 	const initialComputedWidthRef = useRef( null );
 	const FRAME_TRANSITION = { type: 'tween', duration: isResizing ? 0 : 0.5 };
+	const frameRef = useRef( null );
+
+	// Remember frame dimensions on initial render.
+	useEffect( () => {
+		const { offsetWidth, offsetHeight } = frameRef.current.resizable;
+		initialComputedWidthRef.current = offsetWidth;
+		initialAspectRatioRef.current = offsetWidth / offsetHeight;
+	}, [] );
+
+	const handleResizeStart = ( _event, _direction, ref ) => {
+		// Remember the starting width so we don't have to get `ref.offsetWidth` on
+		// every resize event thereafter, which will cause layout thrashing.
+		setStartingWidth( ref.offsetWidth );
+		setIsResizing( true );
+	};
 
 	// Calculate the frame size based on the window width as its resized.
-	const handleResize = ( event, direction, ref ) => {
-		const updatedWidth = ref.offsetWidth;
-		const updatedHeight = ref.offsetHeight;
+	const handleResize = ( _event, _direction, _ref, delta ) => {
```

Comment:
The split across separate functions makes the logic much more clear:

- initial `useEffect` to calculate the "first-render" width and aspect ratio
- when the resizing starts, get the starting resize width and set `isResizing` to `true`
- the `onResize` callback only takes care of updating the resizable box size, and setting `isOversized`
- when the resizing stops, check if we need to snap back to initial position or to switch to edit mode, and reset all the flags appropriately.